src/go/k8s/controllers/redpanda/cluster_controller_configuration.go

```diff
@@ -0,0 +1,275 @@
+	if err != nil {
+		return fmt.Errorf("could not get centralized configuration schema from cluster %s/%s: %w", redpandaCluster.Namespace, redpandaCluster.Name, err)
+	}
+	clusterConfig, err := adminAPI.Config()
+	if err != nil {
+		return fmt.Errorf("could not get current centralized configuration from cluster %s/%s: %w", redpandaCluster.Namespace, redpandaCluster.Name, err)
+	}
+
+	status, err := adminAPI.ClusterConfigStatus()
+	if err != nil {
+		return fmt.Errorf("could not get current centralized configuration status from cluster %s/%s: %w", redpandaCluster.Namespace, redpandaCluster.Name, err)
+	}
+	var invalidProperties []string
+	for i := range status {
+		invalidProperties = append(invalidProperties, status[i].Invalid...)
+		invalidProperties = append(invalidProperties, status[i].Unknown...)
+	}
+
+	lastAppliedConfiguration, _, err := configMapResource.GetLastAppliedConfigurationFromCluster(ctx)
+	if err != nil {
+		return err
+	}
+
+	patch := configuration.FourWayMerge(config.ClusterConfiguration, clusterConfig, lastAppliedConfiguration, invalidProperties)
+	if !patch.Empty() {
+		log.Info("Applying patch to the cluster configuration", "patch", patch.String())
+		_, err := adminAPI.PatchClusterConfig(patch.Upsert, patch.Remove)
+		if err != nil {
+			return fmt.Errorf("could not patch centralized configuration on cluster %s/%s: %w", redpandaCluster.Namespace, redpandaCluster.Name, err)
+		}
+	}
+
+	// TODO a failure and restart here may lead to inconsistency if the user changes the CR in the meantime (e.g. removing a field),
+	// since we applied a config to the cluster but did not store the information anywhere else.
+	// A possible fix is doing a two-phase commit (first stage commit on configmap, then apply it to the cluster, with possibility to recover on failure),
+	// but it seems overkill given that the case is rare and requires cooperation from the user.
+
+	hash, err := config.GetCentralizedConfigurationHash(schema)
+	if err != nil {
+		return fmt.Errorf("could not compute hash of the new configuration for cluster %s/%s: %w", redpandaCluster.Namespace, redpandaCluster.Name, err)
+	}
+
+	oldHash, err := statefulSetResource.GetCentralizedConfigurationHashFromCluster(ctx)
+	if err != nil {
+		return err
+	}
+
+	if oldHash == "" {
+		// Annotation not yet set on the statefulset (e.g. first time we change config).
+		// We check a diff against last applied configuration to avoid triggering a restart when not needed.
+		prevConfig := *config
+		prevConfig.ClusterConfiguration = lastAppliedConfiguration
+		oldHash, err = prevConfig.GetCentralizedConfigurationHash(schema)
+		if err != nil {
+			return err
+		}
+	}
+
+	if oldHash != hash {
+		// Definitely needs restart
+		if err := statefulSetResource.SetCentralizedConfigurationHashInCluster(ctx, hash); err != nil {
+			return fmt.Errorf("could not update config hash on statefulset %s/%s: %w", redpandaCluster.Namespace, redpandaCluster.Name, err)
+		}
+	}
```

Comment:
I think this has been addressed in the core with validation. Once the cluster is started for the very first time, the only way to change the centralized configuration is via admin API (.bootstrap.yaml file has no effect), so the cluster will never accept a configuration that could lead to a bootstrap crash.. or at least, it would be a bug to fix immediately..